Risk Trajectory Split 4 : Non-interpolated (Static) trajectories

spjuhel opened the pull request:
As the Risk trajectory PR is too substantial, this is a fourth split.

This PR builds on PRs #1197 and #1199, introduces `RiskTrajectory`, the abstract basis for trajectory objects, and the concrete class `StaticRiskTrajectory` which aims at facilitating the computation of risk metrics over multiple `Snapshot` objects.

The code is split between a high level interface `StaticRiskTrajectory` and a low level computing class `CalcRiskMetricsPoint`.

`RiskTrajectory`
=================

The rationale behind this class is to hold the common code for any type of risk trajectory:
- a collection of Snapshot
- an optional risk discount rate, and associated methods to compute NPV.
- (abstract) access to risk metrics

`CalcRiskMetricPoint`
=================

The rationale behind this class is to handle the computation and "data formatting" of different risk metrics over a collection of `Snapshot`, following an `ImpactComputationStrategy` (see PR #1199).  In this case, without any interpolation, thus it is essentially a wrapper for multiple calls to `ImpactCalc` or equivalent in `ImpactComputationStrategy`.

The class contains:
- `@lazy_property` properties, dictating how the foundational elements for risk metric should be computed (see below)
- `calc_<risk_metric>` methods (which are the ones called by the `StaticRiskTrajectory` object): these make use of the above and put them in nice and tidy dataframes.

`@lazy_property` explained
----------------------------------------

We want to avoid recomputing when unnecessary, thus we have computed metrics and "impact data" as `@lazy_property`, which cache the results of the computation. For instance, calling `static_traj.per_date_aai` the first time computes the different required `Impact` objects, stores them, and then the `.aai_agg` for each `Impact` and also store them.

A subsequent call to `.per_date_aai` return the previously stored value.
In addition, a subsequent call to `.per_date_eai` will reuse computed `Impact` objects. 

The `@lazy_property` decorator implements this memoization flow: it returns the cached `_metric` attribute if available. If not, it computes the value, stores it in `_metric`, and returns the result.

Attributes that would change metric results (such as the computation strategy or the return periods to compute) make use of `@property` to reset the cached data when their value is changed by calling `_reset_impact_data`.

`StaticRiskTrajectory`
=================

This class is the interface with the user. It receives the list of snapshot to compute metrics for, and gives access to computed risk metrics and plots. It also handles the computation of net present values if a risk discount rate is provided.
It also uses a memoization flow for the "final" results.


### PR Author Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] Correct target branch selected (if unsure, select `develop`)
- [ ] Descriptive pull request title added
- [ ] Source branch up-to-date with target branch
- [ ] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [ ] [Tests][testing] updated
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/development/Guide_Review.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/development/Guide_continuous_integration_GitHub_actions.html
[linter]: https://climada-python.readthedocs.io/en/latest/development/Guide_continuous_integration_GitHub_actions.html#static-code-analysis

spjuhel committed:
cherrypick from risk_traj

spjuhel committed:
Merge branch 'feature/interpolation-strategies' into feature/static-risk-traj

spjuhel committed:
Merge branch 'feature/impact-computation-strategies' into feature/static-risk-traj

spjuhel committed:
cherry pick, renaming

spjuhel committed:
cherry picks __init__

spjuhel committed:
Merge branch 'feature/impact-computation-strategies' into feature/static-risk-traj

spjuhel committed:
namespace fixes

spjuhel committed:
cherry picks dataframe handling

spjuhel committed:
Introduces on/off option for caching

spjuhel committed:
removes redondant code

spjuhel committed:
Clarifies docstring

spjuhel committed:
Cherry picks tests

spjuhel committed:
Initial data

spjuhel committed:
Merge branch 'feature/common_test_fixtures' into feature/static-risk-traj

spjuhel committed:
cleanups test

spjuhel committed:
cleansup

spjuhel committed:
Merge branch 'feature/common_test_fixtures' into feature/static-risk-traj

spjuhel committed:
Pylint compliance

spjuhel committed:
Fixes type hints

spjuhel committed:
Merge branch 'feature/snapshots' into feature/static-risk-traj

spjuhel committed:
Merge branch 'feature/impact-computation-strategies' into feature/static-risk-traj

spjuhel committed:
API references files

spjuhel committed:
Merge branch 'feature/snapshots' into feature/static-risk-traj

spjuhel committed:
Merge branch 'feature/impact-computation-strategies' into feature/static-risk-traj

spjuhel committed:
Wrong merge somewhere pbly

spjuhel committed:
cleanup from wrong merge

spjuhel committed:
fixes type hint annoyance

spjuhel committed:
Shifts tests to pytest

spjuhel committed:
cleanup in tests

spjuhel committed:
Merge branch 'feature/snapshots' into feature/static-risk-traj

spjuhel committed:
Changes remaining test to pytest

spjuhel committed:
Fixes inconsistencies with all_groups_name

spjuhel committed:
idid for tests

spjuhel committed:
Fixes from integration tests

spjuhel committed:
Merge branch 'feature/impact-computation-strategies' into feature/static-risk-traj

spjuhel committed:
Fix discount rates

spjuhel committed:
Merge remote-tracking branch 'origin/develop' into feature/static-risk-traj

spjuhel committed:
Merge branch 'develop' into feature/static-risk-traj

spjuhel committed:
Merge branch 'feature/snapshots' into feature/static-risk-traj

spjuhel committed:
Merge branch 'feature/impact-computation-strategies' into feature/static-risk-traj

spjuhel committed:
Merge remote-tracking branch 'origin/develop' into feature/static-risk-traj

spjuhel committed:
Updates tests

chahank commented on climada/test/common_test_fixtures.py:
```diff
@@ -0,0 +1,205 @@
+    )
+
+
+def reusable_intensity_mat(max_intensity=HAZARD_MAX_INTENSITY):
+    # Choosen such that:
+    # - 1st event has 0 intensity
+    # - 2nd event has max intensity in first exposure point (defaulting to 0 value)
+    # - 3rd event has 1/2* of max intensity in second centroid
+    # - 4th event has 1/4* of max intensity everywhere
+    # *: So that you can double intensity of the hazard and expect double impacts
+    return csr_matrix(
```
What would happen with negative intensities? Probably easy to test using a negative intensity factor. Although then the reusable impact function set would need adjustments.

spjuhel replied:
I think that is beyond the scope of this PR.

chahank replied:
hmmm when could this be done? Because negative intensities are valid in CLIMADA, and it would be quite problematic to break this without good reason.

spjuhel replied:
I'll have a look.

spjuhel replied:
OK, sorry I got confused actually, my initial comment on "this is beyond this PR" applies.

This PR has absolutely no impact on how negative intensities are handled by CLIMADA, it will work the same as before, we did not modify anything related to impact computation here, everything is just a wrapper.

I do have to check if interpolation cause problems with negative value (but that's split 5 not this one).

What also might change something is the option appraisal refactoring (next big project in line) which does change the `measure` module, and hopefully fixes the problem mentioned by Chris in #1257.

chahank commented on climada/test/common_test_fixtures.py:
```diff
@@ -0,0 +1,205 @@
+                max_intensity / 4,
+                max_intensity / 4,
+                max_intensity / 4,
+            ],
```
Formatting.

spjuhel replied:
?
Black does not complain with me?

chahank replied:
Yes, that is why it is not readable. One needs to turn of Black for good formatting :D

spjuhel replied:
Ok! I disagree with it being not readable, but I guess that is subjective.

Anyway that file disappears in split 5 (it is actually not the proper way to use common fixture). I'll improve it there.

chahank commented on climada/test/test_trajectories.py:
```diff
@@ -0,0 +1,278 @@
+        self.expected_future_imp = ImpactCalc(
+            **self.future_snapshot.impact_calc_data
+        ).impact()
+        # self.group_vector = self.base_snapshot.exposure.gdf[GROUP_ID_COL_NAME]
```
```suggestion
```

spjuhel replied:
Fixed

chahank commented on climada/trajectories/test/test_calc_risk_metrics.py:
```diff
@@ -0,0 +1,346 @@
+                date=str(future_date),
+            ),
+        ],
+        "expected_eai": np.array(
```
Not sure whether this is a good idea to have so many meaningless digits after the comma for test that cannot be verified to be correct.

spjuhel replied:
What do you suggest?

chahank replied:
Can one get results that are reproducible / tractable?

spjuhel replied:
I've now used test data similar to the one in the integration test. 

Notes: I have added a `conftest.py` that contains the common fixtures for `pytest` (they are not to be imported, pytest knows to use the names defined in `conftest.py`)

chahank commented on climada/trajectories/calc_risk_metrics.py:
```diff
@@ -0,0 +1,358 @@
+_CACHE_SETTINGS = {"ENABLE_LAZY_CACHE": False}
+
+
+def lazy_property(method):
```
Does this maybe belong into the utils instead?

chahank replied:
Also, is there a reason to not use [@functools.cached_property(func)](https://docs.python.org/3/library/functools.html#functools.cached_property)

spjuhel replied:
Possibly, I didn't know/find this feature (and wanted to avoid lru_cache).
I'll have a look.

spjuhel replied:
So toggling caching makes the decorator as complex if not more with `cached_property`. I'd avoid making the change.

chahank commented on climada/trajectories/calc_risk_metrics.py:
```diff
@@ -0,0 +1,358 @@
+
+        """
+
+        self._reset_impact_data()
```
Why is this needed in the __init__ ?

spjuhel replied:
Otherwise, "_impacts", "_eai_gdf", ... are not existing attribute member.

We could just set this attribute to None, but then one needs to carefully update `_reset_impact_data` if changes are made to the computed data attributes.

For clarity, I changed the method's name to `_init_impact_data`, and have `_reset_impact_data` as an alias.

chahank replied:
Good for me.

chahank commented on climada/trajectories/calc_risk_metrics.py:
```diff
@@ -0,0 +1,358 @@
+        except ValueError as exc:
+            error_message = str(exc).lower()
+            if "need at least one array to concatenate" in error_message:
+                self._group_id = np.array([])
```
This seems to not generate actionable error messages. How should the user know what to do after getting this error?

spjuhel replied:
This **catches** the error caused by the absence of a group_id column in all snapshots and acts accordingly (defining the group_id as an empty array). No error are raised here. 
We check the content of the error to only handle that case.

I did forget to raise if this is not the case. Will fix this.

spjuhel replied:
Fixed

chahank commented on climada/trajectories/calc_risk_metrics.py:
```diff
@@ -0,0 +1,358 @@
+    @impact_computation_strategy.setter
+    def impact_computation_strategy(self, value, /):
+        if not isinstance(value, ImpactComputationStrategy):
+            raise ValueError("Not an impact computation strategy")
```
Not a clear message for the user. Please write error messages that allow the user to solve their problem.

spjuhel replied:
Hopefully fixed.

chahank commented on climada/trajectories/calc_risk_metrics.py:
```diff
@@ -0,0 +1,358 @@
+    def apply_measure(self, measure: Measure) -> "CalcRiskMetricsPoints":
+        """Creates a new `CalcRiskMetricsPoints` object with a measure.
+
+        The given measure is applied to both snapshot of the risk period.
```
That is quite unclear.

spjuhel replied:
Does this clarify:

```
        """Creates a new `CalcRiskMetricsPoints` object by applying the effects
        of the given measure.

        The effects of the measure are applied to all the snapshots contained
        in the initial `CalcRiskMetricsPoints` and a new `CalcRiskMetricsPoints`
        containing the modified snapshots is returned.

        Parameters
        ----------
        measure : Measure
            The measure to apply.

        Returns
        -------

        CalcRiskMetricsPoints
            The risk period with given measure applied.

        """
```

chahank commented on climada/trajectories/calc_risk_metrics.py:
```diff
@@ -0,0 +1,358 @@
+        -------
+
+        CalcRiskPeriod
+            The risk period with given measure applied.
```
Do you mean a `CalcRiskMetricsPoints` ?

spjuhel replied:
Legacy docstring from before, corrected

chahank commented on climada/trajectories/static_trajectory.py:
```diff
@@ -0,0 +1,315 @@
+    This class computes risk metrics over a series of snapshots,
+    optionally applying risk discounting. It does not interpolate risk
+    between the snapshot and only provides results for each snapshot.
+
```
This could be made clearer.

spjuhel replied:
How/What is unclear?

chahank commented on climada/trajectories/static_trajectory.py:
```diff
@@ -0,0 +1,315 @@
+        RETURN_PERIOD_METRIC_NAME,
+        AAI_PER_GROUP_METRIC_NAME,
+    ]
+    """Class variable listing the risk metrics that can be computed.
```
Arent't the full impact matrices computed and accessible?

spjuhel replied:
Not at that high-level class.
This class's purpose is to offer access to EAI and all its aggregations in tidy DataFrames, following a consistent framework.

I'm not sure, I see the point of giving access to impact matrices here.
Users can use `ImpactCalc` / `Snapshot` directly if they want that level of information.

chahank replied:
I see. But this matters for the non-static trajectories right? Because interpolation of impact matrics is not accessible with ImpactCacl/Snapshot.

spjuhel replied:
Yes? But again, the idea of this module is not really to give access to interpolated risk matrices (because we can't do that for all dimensions of risk, only for exposure really)

The `CalcRiskMetricsPeriod` object defined in split 5 takes care of the interpolation process, and does hold the various arrays of interpolated impact matrices between the pair of snapshots it is defined for, arrays needed to compute the risk metrics. As well as generic methods/functions to compute these arrays.

Again, you cannot have interpolated impact matrices between two snapshots, only between some of the edges of the risk cubes (the ones that share the same hazard), and I would rather not make a direct interface to that.

chahank commented on climada/trajectories/static_trajectory.py:
```diff
@@ -0,0 +1,315 @@
+            Defaults to `DEFAULT_RP` ([20, 50, 100]).
+        risk_disc_rates: DiscRates, optional
+            The discount rate to apply to future risk. Defaults to None.
+        impact_computation_strategy: ImpactComputationStrategy, optional
+            The method used to calculate the impact from the (Haz,Exp,Vul)
+            of the two snapshots. Defaults to :class:`ImpactCalcComputation`.
+
```
What is meant by the "two snapshots". The input is a list of snapshots.

spjuhel replied:
Legacy description from when there was only interpolated trajectories (and pairs of snapshots)
Corrected.

chahank commented on climada/trajectories/static_trajectory.py:
```diff
@@ -0,0 +1,315 @@
+    @impact_computation_strategy.setter
+    def impact_computation_strategy(self, value, /):
+        if not isinstance(value, ImpactComputationStrategy):
+            raise ValueError("Not an interpolation strategy")
```
Unclear error message for the user. This is likely raised at the init of the object.

spjuhel replied:
Improved

chahank commented on climada/trajectories/static_trajectory.py:
```diff
@@ -0,0 +1,315 @@
+        if COORD_ID_COL_NAME in tmp.columns:
+            tmp = tmp.set_index([COORD_ID_COL_NAME], append=True)
+
+        # When more than 2 snapshots, there might be duplicated rows, we need to remove them.
+        # Should not be the case in static trajectory, but in any case we really don't want
+        # duplicated rows, which would mess up some dataframe manipulation down the road.
```
Should this not be made clear to the user in case it happens? Since some data is arbitrarily discarded.

spjuhel replied:
I added a note in the docstring and a warning.

chahank commented on climada/trajectories/static_trajectory.py:
```diff
@@ -0,0 +1,315 @@
+        Notes
+        -----
+
+        This computation may become quite expensive for big areas with high resolution.
```
```suggestion
        This computation may become quite expensive for exposures with many points (e.g., big areas with high resolution).
```

spjuhel replied:
Suggestion implemented

chahank commented on climada/trajectories/trajectory.py:
```diff
@@ -0,0 +1,261 @@
+    @risk_disc_rates.setter
+    def risk_disc_rates(self, value, /):
+        if value is not None and not isinstance(value, (DiscRates)):
+            raise ValueError("Risk discount needs to be a `DiscRates` object.")
```
What is Risk discount?

spjuhel replied:
Changed to "The discount rate applied to risk values"

chahank commented on climada/util/dataframe_handling.py:
```diff
@@ -0,0 +1,63 @@
+import pandas as pd
+
+
+def reorder_dataframe_columns(
```
A simple test would be useful as this has the potential to screw up outputs if pandas changes the way columns are indexed / selected.

spjuhel replied:
Test added

spjuhel committed:
Implements Chahan's comments

spjuhel committed:
Fixes tests

spjuhel committed:
Merge remote-tracking branch 'origin/develop' into feature/static-risk-traj

spjuhel committed:
Adds conftest, makes test more tractable

spjuhel committed:
adds test for dataframe handling

spjuhel commented:
@chahank 
I am pretty sure I addressed all comments, see the one on negative intensities and resolve if you are OK.

The failing tests are related to #1263 #1264 (and #1081) and are being addressed, and not related to this PR.

I'll merge once you greenlight me :)

spjuhel committed:
Merge remote-tracking branch 'origin/develop' into feature/static-risk-traj
